### DISCUSSION TITLE
Normalization methods

### DISCUSSION THREAD

**Discussion by mychelmorais** [fgcz/prolfqua#69] [Q&A] [answered]
Can I implement other normalization approaches to peptide data other than robust scaling, such as quantile, or vsn?

I have tried implement it through this code, but I am not managing to find the right function:

```
sol_r6 <- sol_r6$get_Transformer()$log2()$vsn()$lfq
sol_r6 <- sol_r6$get_Transformer()$log2()$quantiles()$lfq

Error: attempt to apply non-function
```

**Comment by wolski** [comment]
Yes, see the example code in:

https://fgcz.github.io/prolfqua/reference/LFQDataTransformer.html

specifically the last few lines:

```
if([require](https://rdrr.io/r/base/library.html)(["vsn"](http://www.r-project.org/))){
 res <- lfqTrans$intensity_matrix( .func = vsn::[justvsn](https://rdrr.io/pkg/vsn/man/justvsn.html))
 res$lfq$get_Plotter()$intensity_distribution_density()
 res$get_scales()
}
```

The transformer class has two functions:

`intensity_matrix`
and `intensity_array`

`intensity_matrix` takes as an argument a method working and returning matrices, e.g., `justvsn`  or `normalize.quantiles`
and applies them to the response (abundances).

 `intensity_array` takes as an argument any function working with arrays e.g. `sqrt`, `log2`, and applies them to the response (abundances).

**Reply by mychelmorais** [reply]
Thanks!! I've tried the vsn and worked fine, but with normalize.quantiles I am getting the following error...

This is my code:

```
# normalize with quantiles
soluble_r6$config$table$is_response_transformed
soluble_r6 <- soluble_r6$get_Transformer()
soluble_r6 <- soluble_r6$intensity_array(log2)
soluble_r6 <- soluble_r6$intensity_matrix(
  preprocessCore::normalize.quantiles, force = TRUE)$lfq
plot <- soluble_r6$get_Plotter()
plot$intensity_distribution_density()+theme(legend.position = "none")
```

Error in `dplyr::bind_cols()`:
! Can't recycle `..1` (size 0) to match `..2` (size 30502).
Backtrace:
 1. soluble_r6$intensity_matrix(...)
 2. prolfqua::apply_to_response_matrix(...)
 3. prolfqua::response_matrix_as_tibble(mat, colname, config, data)
 4. dplyr::bind_cols(tibble::tibble(`:=`("row.names", rownames(pdata))), tibble::as_tibble(pdata))
Error in dplyr::bind_cols(tibble::tibble(`:=`("row.names", rownames(pdata))),  :

**Comment by wolski** [comment, marked as answer]
Can you please try:
```
quant <- function(y){
  ynorm <- preprocessCore::normalize.quantiles(y)
  rownames(ynorm) <- rownames(y)
  colnames(ynorm) <- colnames(y)
  return(ynorm)
 }
 res <- lfqTrans$intensity_matrix( .func = quant)
```

It seems that `normalize.quantiles` drops row and colnames.
Also you need to log first.

**Reply by mychelmorais** [reply]
Yes, it worked! Thanks!!

**Comment by mychelmorais** [comment]
Is it possible to implement other approaches, such as MsCoreUtils methods ("center.median", "center.mean")?

**Comment by wolski** [comment]
Easily, however, they have no practical relevance, which is also the reason why I do not implemented them in prolfqua

```
center.median <- function(mat){
# Function to scale columns by median
scale_by_median <- function(x) {
  median_val <- median(x)
  scaled_col <- x / median_val
  return(scaled_col)
}

# Apply the function to each column of the matrix
scaled_mat <- apply(mat, 2, scale_by_median)
return(scaled_mat)
} 

# and then you just
soluble_r6$intensity_matrix(center.median)
```
